File: src/engine/StatePatch.ts
import { InkObject } from "./Object";
import { Container } from "./Container";

export class StatePatch {
  get globals() {
    return this._globals;
  }
  get changedVariables() {
    return this._changedVariables;
  }
  get visitCounts() {
    return this._visitCounts;
  }
  get turnIndices() {
    return this._turnIndices;
  }

  constructor();
  constructor(toCopy: StatePatch | null);
  constructor() {
    if (arguments.length === 1 && arguments[0] !== null) {
      let toCopy = arguments[0] as StatePatch;
      this._globals = new Map(toCopy._globals);
      this._changedVariables = new Set(toCopy._changedVariables);
      this._visitCounts = new Map(toCopy._visitCounts);
      this._turnIndices = new Map(toCopy._turnIndices);
    } else {
      this._globals = new Map();
      this._changedVariables = new Set();
      this._visitCounts = new Map();
      this._turnIndices = new Map();
    }
  }

  public TryGetGlobal(name: string | null, /* out */ value: InkObject | null) {
    if (name !== null && this._globals.has(name)) {
      return { result: this._globals.get(name), exists: true };
    }

    return { result: value, exists: false };
  }

  public SetGlobal(name: string, value: InkObject) {
    this._globals.set(name, value);
  }

  public AddChangedVariable(name: string) {
    return this._changedVariables.add(name);
  }

  public TryGetVisitCount(container: Container, /* out */ count: number) {
    if (this._visitCounts.has(container)) {
      return { result: this._visitCounts.get(container), exists: true };
    }

    return { result: count, exists: false };
  }

  public SetVisitCount(container: Container, count: number) {
    this._visitCounts.set(container, count);
  }

  public SetTurnIndex(container: Container, index: number) {
    this._turnIndices.set(container, index);
  }

  public TryGetTurnIndex(container: Container, /* out */ index: number) {
    if (this._turnIndices.has(container)) {
      return { result: this._turnIndices.get(container), exists: true };
    }

    return { result: index, exists: false };
  }

  private _globals: Map<string, InkObject>;
  private _changedVariables: Set<string> = new Set();
  private _visitCounts: Map<Container, number> = new Map();
  private _turnIndices: Map<Container, number> = new Map();
}

File: ink_dart/lib/src/engine/state_patch.dart
import 'ink_object.dart';
import 'container.dart';

/// Temporary storage for state changes during background saves.
///
/// Allows the story to continue progressing while a background
/// save is in progress without corrupting the saved state.
class StatePatch {
  final Map<String, InkObject> _globals = {};
  final Map<Container, int> _visitCounts = {};
  final Map<Container, int> _turnIndices = {};
  final Set<String> _changedVariables = {};

  StatePatch([StatePatch? toCopy]) {
    if (toCopy != null) {
      _globals.addAll(toCopy._globals);
      _visitCounts.addAll(toCopy._visitCounts);
      _turnIndices.addAll(toCopy._turnIndices);
      _changedVariables.addAll(toCopy._changedVariables);
    }
  }

  /// All globals in the patch.
  Map<String, InkObject> get globals => _globals;

  /// All visit counts in the patch.
  Map<Container, int> get visitCounts => _visitCounts;

  /// All turn indices in the patch.
  Map<Container, int> get turnIndices => _turnIndices;

  /// Changed variable names.
  Set<String> get changedVariables => _changedVariables;

  /// Try to get a global variable from the patch.
  ({bool exists, InkObject? result}) tryGetGlobal(String? name, InkObject? defaultValue) {
    if (name != null && _globals.containsKey(name)) {
      return (exists: true, result: _globals[name]);
    }
    return (exists: false, result: defaultValue);
  }

  /// Set a global variable in the patch.
  void setGlobal(String name, InkObject value) {
    _globals[name] = value;
  }

  /// Add a variable to the changed set.
  void addChangedVariable(String name) {
    _changedVariables.add(name);
  }

  /// Try to get visit count for a container.
  ({bool exists, int? result}) tryGetVisitCount(Container container, int defaultValue) {
    if (_visitCounts.containsKey(container)) {
      return (exists: true, result: _visitCounts[container]);
    }
    return (exists: false, result: defaultValue);
  }

  /// Set visit count for a container.
  void setVisitCount(Container container, int count) {
    _visitCounts[container] = count;
  }

  /// Try to get turn index for a container.
  ({bool exists, int? result}) tryGetTurnIndex(Container container, int defaultValue) {
    if (_turnIndices.containsKey(container)) {
      return (exists: true, result: _turnIndices[container]);
    }
    return (exists: false, result: defaultValue);
  }

  /// Set turn index for a container.
  void setTurnIndex(Container container, int index) {
    _turnIndices[container] = index;
  }
}
